src/js/components/Breadcrumb.js

```diff
@@ -0,0 +1,295 @@
+import classNames from 'classnames';
+import DeepEqual from 'deep-equal';
+import {Link} from 'react-router';
+import React from 'react';
+import ReactDOM from 'react-dom';
+
+import IconChevron from './icons/IconChevron';
+import StringUtil from '../utils/StringUtil';
+
+const COLLAPSE_BUFFER = 12;
+const LAST_ITEM_OFFSET = 150; // Difference between scrollWidth and outerWidth
+const PADDED_ICON_WIDTH = 38; // Width of icon + padding
+
+const {PropTypes} = React;
+
+class Breadcrumb extends React.Component {
+  constructor() {
+    super();
+
+    this.displayName = 'Breadcrumb';
+
+    this.state = {
+      collapsed: false,
+      availableWidth: null,
+      expandedWidth: null
+    };
+    this.handleResize = this.handleResize.bind(this);
+  }
+
+  componentDidMount() {
+    if (global.window != null) {
+      window.addEventListener('resize', this.handleResize);
+      window.addEventListener('focus', this.handleResize);
+    }
+  }
+
+  componentDidUpdate() {
+    this.updateDimensions();
+  }
+
+  componentWillUnmount() {
+    if (global.window != null) {
+      window.removeEventListener('resize', this.handleResize);
+      window.removeEventListener('focus', this.handleResize);
+    }
+  }
+
+  shouldComponentUpdate(nextProps, nextState) {
+    if (nextState.availableWidth == null || nextState.expandedWidth == null) {
+      this.updateDimensions();
+
+      return false;
+    }
+    if (this.state.collapsed !== nextState.collapsed) {
+      return true;
+    }
+
+    return !DeepEqual(this.props, nextProps);
+  }
+
+  handleResize() {
+    let availableWidth = this.getAvailableWidth();
+    this.setState({
+      availableWidth,
+      collapsed: this.shouldCollapse(availableWidth, this.state.expandedWidth)
+    });
+  }
+
+  updateDimensions() {
+    let availableWidth = this.getAvailableWidth();
+    let expandedWidth = this.getExpandedWidth();
+
+    this.setState({
+      availableWidth,
+      expandedWidth,
+      collapsed: this.shouldCollapse(availableWidth, expandedWidth)
+    });
+  }
+
+  getAvailableWidth() {
+    return ReactDOM.findDOMNode(this).offsetWidth;
+  }
+
+  getWidthFromExpandedItem(item) {
+    return item.offsetWidth + PADDED_ICON_WIDTH;
+  }
+
+  getWidthFromCollapsedItem(item) {
+    let link = item.children[0];
+    let textWidth = link.scrollWidth - link.offsetWidth;
+
+    return textWidth + PADDED_ICON_WIDTH;
+  }
+
+  getExpandedWidth() {
+    // array/splat casts NodeList to array
+    let listItems = [...ReactDOM.findDOMNode(this).children]
+      .filter(function (_, i) {
```

Comment:
Not conveniently because we use `listItems.length` in the `map` function. We could of course get round that with some math and put everything inside a `reduce` but I think it will make it less obvious. The `listItems.length` will rarely be > 10 I'd imagine so there isn't any performance concerns.